core/src/read_write_account_set.rs

```diff
@@ -0,0 +1,354 @@
+use {
+    solana_sdk::{
+        message::{SanitizedMessage, VersionedMessage},
+        pubkey::Pubkey,
+    },
+    std::collections::HashSet,
+};
+
+/// Wrapper struct to check account locks for a batch of transactions.
+#[derive(Debug, Default)]
+pub struct ReadWriteAccountSet {
+    /// Set of accounts that are locked for read
+    read_set: HashSet<Pubkey>,
+    /// Set of accounts that are locked for write
+    write_set: HashSet<Pubkey>,
+}
+
+impl ReadWriteAccountSet {
+    /// Check static account locks for a transaction message.
+    pub fn check_static_account_locks(&self, message: &VersionedMessage) -> bool {
```

Comment:
Still need to look into further, but is wondering if over engineered to separately checking static and lookup accounts, if transactions are required to be fully deserailzied and sanitized up front?

Reply:
Might be over-engineered. I can re-run my test-suite with and without this separation to give an idea of speed up.

My test has all static keys, so it'll really only be measuring the additional checks and allocations done in sanitization. In more realistic scenarios this would also avoid looking up lookup tables *if* they are used in the tx.

Reply:
| mint% | static check before sanitization | sanitized check only |
| --: | --: | --: |
| 0 | 30336 | 29120 |
| 50 | 20352 | 14592 |
| 75 | 34624 | 11328 |
| 85 | 33088 | 9600 |
| 90 | 30877 | 7744 |
| 95 | 25328 | 5734 |
| 99 | 13398 | 3458 |
| 99.9 | 8312 | 1980 |
| 100 | 7785 | 1842 |

Sanitization is expensive - best to avoid doing it multiple times. We might be able to do something better by keeping sanitized txs once. IMO better to do that in a separate change, as is the static-check first is helping in a significant way.

Reply:
Ah, that makes sense. I was thinking that transaction probably will have to be sanitized way up front, even before sigverify. If that'd be the case, checking static accounts separately might not be necessary. But good to leave this to next PRs.

Reply:
Yeah if we're doing sanitization work up front (I think we should), then this separation goes away and life becomes easier.